### ISSUE TITLE
Lookup is slow?

### ISSUE THREAD

**Issue by protist** [scheibler/khard#103]
I'm finding khard to be a bit slow when looking up addresses. I originally noticed this in mutt, but I've tested it directly a few times.
```
$ time khard email foo
Found no email addresses

================
CPU	99%
user	1.383
system	0.057
total	1.442
```
Is that normal? Should khard be taking ~1.4 seconds for each lookup? This is more of an issue when I send emails to several people. The cumulative wait adds up!

**Comment by scheibler** [comment]
A certain delay is not preventable due to module initializations and command line parsing
(0.3-0.4 seconds at least). The rest depends on the size of your address books. You may try the
--search-in-source-files option to speed up search for email addresses (probably saves you 0.3-0.5
seconds).

```
khard email --search-in-source-files foo bar
```

**Comment by protist** [comment]
I ran a test comparing both command, for ten runs each.
* `khard email foo` takes an average of 1.43 seconds.
* `khard email --search-in-source-files foo` takes an average of 1.18 seconds.

That's a saving of ~14%.

FWIW I have 158 contacts. 148 have an average of 11.4 lines in their `.vcf` file. The final 10 have an average of 936 lines; these are the cards with associated photos.

I know that khard is extremely sophisticated, but I was just surprised as its speed compared to simpler text parsing scripts. For example, as an extremely blunt tool, I tested the following in my contacts (`*.vcf`) directory, where `foo` is actually someone's name.

```
<$(grep foo -rl) awk -F: '$1 ~ /^EMAIL;/ {print $2}'
```
This took an average of 0.0098 seconds to run.

**Comment by GSI** [comment]
@protist Use this to pre-generate a file containing the addresses. Running it once a day should suffice in most cases:

`(khard --skip-unparsable email | grep -v "^searching" | sed "s#  \+#\t#g" | sed "s#\(.*\)\t\(.*\)\t\(.*\)\t#\3
\t\1\t\2#g" > ~/.mutt/address_index_from_vcards) || echo 'ERROR: Aborting.'`


Then **mutt** will present you with immediate results with this query command:

    set query_command = "echo 'Searching ...' ; grep -i %s'.*pref' ~/.mutt/address_index_from_vcards || grep -i %s ~/.mutt/address_index_from_vcards"

I too am surprised about khard's slow lookups and suspect there is some "bug" that can be tracked & resolved - but even if khards becomes quicker it will never outperform a simple `grep`.

**Comment by lucc** [comment]
I am having the same problem with slow lookups. My address book is considerably bigger (700-1200 entries) but I too believe that some optimization can be done in khard. I started a PR to prepare for some bigger work on speed improvment a while ago but currently there is no new input on the discussion. If you are interested I urge you to jump the discussion and help us move things forwards. Its #92.

In the meanwhile I am using a simple wrapper that uses the idea @GSI mentioned but instead of a cron job I use a makefile and can hence be sure that the data I get is always up to date. I put it up as a gist for now as it is not yet mature enough for its own repo: https://gist.github.com/lucc/89437c04af71ec4e818b1c0c31c01f69
(The name comes from ldbd which I used before but that does not offer caching and hence inherits the slowness of all its providers: http://www.spinnaker.de/lbdb/ )

**Comment by protist** [comment]
@GSI Fantastic! FWIW even building that database once takes about as long as the original khard query (~1.447s). The lookup is obviously blazingly fast.

I'm not sure if it's somehow specific to my system (latest `khard`, 95ef4cf), but I found a few bugs in it.
* If there are spaces in names (e.g. between first and last names), the script converts these into tabs. Hence mutt only displays the first name.
* It also contains the khard header `E-Mail 	Name	Type`, which is revealed if you try to search in mutt with `Name`.
* I removed the `grep -v "^searching"` part, because it wasn't in the output on my system.
* I decided to rewrite in awk for simplicity.
```
khard --skip-unparsable email | awk -F'\t' -vOFS='\t' 'NR > 1 {gsub(/  +/, "\t"); print $3,$1,$2}' > ~/.mutt/address_index_from_vcards
```
(Also, I didn't mean to imply that khard could approach the speed of simple grep or awk, but was just trying to compare the relative speed on my system.)

Finally, I agree that fixes would preferably be in khard itself, as this does not help for general querying on the command line (`khard show`, etc.), and the lookup within mutt is not quite as powerful. For example, I have several entries, `Chris Foo`, `Chris Bar`, etc. With vanilla khard, I can select the former with `chrisf`, or `ch f`, etc.

**Comment by protist** [comment]
This is probably not consistent with how khard does these lookups, but I further refined the workaround lookup code. It now allows you to grep using parts of names and ignoring spaces. Each group of letters that you type is considered to be the start of a name component. Hence, you can search successfully using `chris foo`, `chrisfoo`, `chrf`, but *not* `hris oo`, etc.

I also made it search only in names, not email addresses. (You can revert this by removing `$2 ~ ` in the `awk` block.)

To avoid escaping hell, I made this a separate script (e.g. as `mutt_query_command`). Save it in your $PATH, and put `set query_command = "mutt_query_command %s"` into `muttrc`.

```
#!/bin/sh
echo 'Searching...'
non_seq="$(<<<"$@" sed -r 's/(.)/\1([^ ]* )*/g')"
awk -F'\t' -vIGNORECASE=1 -vVAR="$non_seq" '$2 ~ VAR' ~/.mutt/address_index_from_vcards
```

**Comment by lucc** [comment]
I did some profiling (I think there is some thread here where we did that before) and found the following: Most time is spend when *parsing* the vcard files. That is also what makes the awk solutions so much faster. They do not parse all the special stuff that might be present in vcard files (quoted printable or base64 encoded strings for example).

I followed these steps to get nice pdf/png output from the profiler: http://stackoverflow.com/a/7693928
The document generated by gprof2dot was much easier to read to get an overview but the other one is grouped by modules. Also pycallgraph was quite slow for me (but that might also be the fault of my big addressbook).

Something interesting was that also for subcommands like addressbooks all the vcard files where parsed. I think that is not necessary. The names can be obtained from the config file alone.

**Comment by scheibler** [comment]
I've analyzed the khard source code manually as well as with cProfile a while ago and encountered
the following two main time killers:

1. Even with an empty address book khard takes about 0.3 seconds to run. That's mostly related to
   module initializations and command line parsing. We can't change the former cause all loaded modules
   are currently required and the time / effort ratio for the latter is much to bad to start at that
   again.

2. The actual vcard parsing. Especially vcards with much binary data are a bit slow (at least slow
   enough, that it matters with several hundred of them). The only optimization I came around so far is
   the --search-in-source-files attribute, which prefilters contacts by a fast but simple grep-like operation.
   But obviously this is not very useful if you search for a rather generic phrase, which is common
   in many contacts.

So yes, we could speed up the addressbooks action and maybe the help text output too but at the
moment I've no idea, how to significantly improve the vcard parsing. Nevertheless I will include the
changes from #92 when I find the time.

**Comment by lucc** [comment]
@scheibler happy to hear your interested in #92. 

I found a another thing which should be easy improve as soon as the code is more object oriented (hence #92): When I profile `khard ls` and `khard ls -a secondary` I get 1235 calls to `khard.carddav_object_from_file` even though secondary only has 441 files (according to `find -type f|wc`). 

That was something which I found out some time ago already and which also motivated me to start the config singleton removal and #92 because there are many places in khards code base where all possible modules and all possible addressbooks/cards are loaded/parsed and so on. I think it is possible to gain something in these cases (this will not help with the plain `khard ls` though). But I need to make the code more object oriented to separate the responsibilities first.

**Comment by GSI** [comment]
@protist I tested your update to the script and can confirm that the output equals mine when run on my address book.

> If there are spaces in names (e.g. between first and last names), the script converts these into tabs. Hence mutt only displays the first name.

Should only happen when there is more than one consecutive space. In your update you do the same but with gsub, or am I missing something?

> It also contains the khard header E-Mail Name Type, which is revealed if you try to search in mutt with Name.

Yes, you either leave that line line in place or add one when running the query (like you do by adding `echo Searching` there; section 6 - External Address Queries in mutt's help).

> I removed the grep -v "^searching" part, because it wasn't in the output on my system.

Guess that line was part of khards output back then.

> Each group of letters that you type is considered to be the start of a name component. 

In my experience the most efficient and precise lookup for names is entering the last 2-3 letters of the first name followed by the first 2-3 letters of the last name. E.g. **ic sc** for Eric Scheibler.

My query_command is tailored to respect email addresses configured as "preferred". This way you avoid manually selecting a specific email address on contacts that have several:

`set query_command = "echo 'Searching ...' ; grep -i %s'.*pref' ~/.mutt/address_index_from_vcards || grep -i %s ~/.mutt/address_index_from_vcards"`

**Comment by protist** [comment]
>Should only happen when there is more than one consecutive space. In your update you do the same but with gsub, or am I missing something?

Erm… I'm not exactly sure now. But I remember with your original version mutt would only display the first name, instead of the whole name, and show the second name as the "type" (i.e. `home` or `work`, etc.). Maybe just some different system/config we have.

>Yes, you either leave that line line in place or add one

Ah yes I see. Thanks for that explanation. Perhaps it's another difference in our systems somehow, but on my system with your code, if I searched for a literal `Name` in mutt, it would bring up the header.

>In my experience the most efficient and precise lookup for names is entering the last 2-3 letters of the first name followed by the first 2-3 letters of the last name. E.g. ic sc for Eric Scheibler.

Ah okay, fair enough. Different use cases then! Cheers.

**Comment by mxmehl** [comment]
I just want to second that the lookup is quite slow on my system too compared to an abook lookup or similar tools – that's quite a pity because khard is quite impressing in most other tasks!

Is there any development in tracking down the bug? Can I (non-programmer) help somehow?

**Comment by polyzen** [comment]
@mxmehl, couldn't find a donation/Paypal link. ;p

**Comment by lucc** [comment]
Hello everybody. I pushed some cleanups to develop and am currently working on #158. For me this already gives some speed improvements:
* in develop `khard abooks` should be faster than in master
* in #158 `khard ls foo` should be faster than in develop
I would be happy about feedback on #158.

**Comment by protist** [comment]
Unfortunately I see no improvement on my system. I did my benchmarking using the following code: `time (for i in {1..10}; do khard ls foo > /dev/null; done)`

For the current HEAD 0d65759: `5.320 s`
For the develop branch gb5c8b2a: `5.290 s`
After applying #158: `5.420 s`

**Comment by lucc** [comment]
@protist what about `khard ls -f foo` ?

**Comment by protist** [comment]
Sorry, I've only had time to look at this now. I'm not even sure what that code does, but in any case, I checked it with `time (for i in {1..10}; do khard ls -f foo > /dev/null; done)`. The current HEAD, develop, and the patch have changed, so I tested with everything as new as possible. The patch doesn't work with the develop head, so I picked one where it does. Unfortunately, there is still no difference anyway.

* For the current HEAD cb430f8: 3.106 s
* For the latest commit 5a4bf45 of the develop branch: 3.196 s
* For commit dccf159 of develop: 3.256 s
* After applying #158 to the commit in the line above: 3.271 s

**Comment by lucc** [comment]
@protist thank you. As #158 is already merged and you also get very similar times it might be moot but: which code do you not understand? Maybe I can write it better or add some comments.

In general I am currently trying to achieve three things (at least in the code, I have also merge quite some tests lately):
1. move code to the class where the responsibility for what it does belongs (code to search several vcards belongs into AddressBook, code to mach one vcard belongs into CardavObject, code to handle several AddressBooks into AddressBookCollection and so on)
2. clean up complicated or badly formatted code and document it
3. try to improve or at least not degrade performance

My hope is that 1 and 2 will help me/us to think about 3.

**General notes and findings about performance**

* When the address book list is filtered on the command line (with -a or -A) *only* the named address books should be parsed/opened/loaded. This could be checked by looking at extra debug output (not currently in the code) and call graphs (see https://github.com/scheibler/khard/issues/103#issuecomment-270330988)
* When search terms are given it depends on `-f`. Without it all vcards from an address book need to be parsed into CarddavObjects, otherwise it could result in less parsing work.
* (**question to all**) Recently I noticed that re.search did take a lot of time in one of my branches (in the callgraph). We should think about the question if regexes are needed or not. Maybe a simple `in` search for strings could yield the same results and be much faster. This also relates to #131.

**Comment by protist** [comment]
@lucc Sorry, I was unclear. I didn't know what the `-f` option did in `khard ls -f foo`.

Regarding the similar times, the code isn't merged in the main branch yet, right? So there is no improvement in processing time. I also just tried the tagged release 0.12.2, and `time (for i in {1..10}; do khard ls -f foo > /dev/null; done)` gives me 2.667 s, which is the quickest yet! (This almost certainly doesn't have that new code.)

**Comment by lucc** [comment]
That is strange because for me origin/HEAD is the same as v0.12.2. So If you get 3.1 and 2.6sec for these the fluctuation on your system seems to be quite high. Try to minimize load (no file copying, no music or video running and so on) and see what your "error of measurement" is.

**Comment by lucc** [comment]
I figured out why the regex took so much time. I opend #167 for that and the move to AddressBookCollection. It gives me a speed improvement in some cases. I would be happy for feedback on #167.

**Comment by lucc** [comment]
#167 is merged into develop so you can check that out. But I think it mostly affects `-f`.

During the work on #167 I noticed that the regex search was a problem in some cases (see above). There are two different functions in pythons `re` module. `re.search` (we use that) which matches anywhere in the string and `re.match` which only matches at the beginning of the string. But I also noted that we currently use the `re` module only for two things:
1. searching case insensitive; we could use `str.casefold()` for that ([why not `str.lower()`?}(https://stackoverflow.com/questions/319426/how-do-i-do-a-case-insensitive-string-comparison#319435)) and then use the `search in string` operator.
2. keeping the order of search terms: currently we match `foo.*bar` for search terms "foo" and "bar" but not `bar.*foo`. This might be an unintended side effect of the implementation. I would also be willing to drop that. If we use `in` we could use `all()` to have a similar search behaviour but order independent. See #131 for more on this point.

In order to be able to argue about the speed of the different search methods, I hacked together this little test script to compare some of the matching functions that we have available:
```py
import re
import timeit
text = open('README.md').read()
foo_plain_search = re.compile('foo', re.IGNORECASE | re.DOTALL).search
foo_star_search = re.compile('.*foo.*', re.IGNORECASE | re.DOTALL).search
foo_ancor_search = re.compile('^.*foo.*$', re.IGNORECASE | re.DOTALL).search
cal_plain_search = re.compile('calendar', re.IGNORECASE | re.DOTALL).search
cal_star_search = re.compile('.*calendar.*', re.IGNORECASE | re.DOTALL).search
cal_ancor_search = re.compile('^.*calendar.*$', re.IGNORECASE | re.DOTALL).search
foo_plain_match = re.compile('.*foo', re.IGNORECASE | re.DOTALL).match
foo_star_match = re.compile('.*foo.*', re.IGNORECASE | re.DOTALL).match
foo_ancor_match = re.compile('^.*foo.*$', re.IGNORECASE | re.DOTALL).match
cal_plain_match = re.compile('.*calendar', re.IGNORECASE | re.DOTALL).match
cal_star_match = re.compile('.*calendar.*', re.IGNORECASE | re.DOTALL).match
cal_ancor_match = re.compile('^.*calendar.*$', re.IGNORECASE | re.DOTALL).match
def foo_in(text): return 'foo' in text
def cal_in(text): return 'calendar' in text
def foo_casefold(text): return 'foo'.casefold() == text.casefold()
def cal_casefold(text): return 'calendar'.casefold() == text.casefold()
n = 10000
results_match = []
results_nomatch = []
for f in [
          'cal_plain_search',
          'cal_star_search',
          'cal_ancor_search',
          'cal_plain_match',
          'cal_star_match',
          'cal_ancor_match',
          'cal_in',
          'cal_casefold',
          ]:
    results_match.append((timeit.timeit('{}(text)'.format(f), globals=globals(), number=n), f))
for f in [
          'foo_plain_search',
          # 'foo_star',
          'foo_ancor_search',
          'foo_plain_match',
          'foo_ancor_match',
          'foo_in',
          'foo_casefold',
          ]:
    results_nomatch.append((timeit.timeit('{}(text)'.format(f), globals=globals(), number=n), f))
for time, name in sorted(results_match):
    print('{:20}\t{}'.format(name, time))
print()
for time, name in sorted(results_nomatch):
    print('{:20}\t{}'.format(name, time))
```
I get these results:
```
cal_in              	0.002246514006401412
cal_star_search     	0.013890016998630017
cal_star_match      	0.013943787009338848
cal_plain_match     	0.014091400007600896
cal_ancor_search    	0.014515878996462561
cal_ancor_match     	0.014880764996632934
cal_plain_search    	0.07135143999767024
cal_casefold        	1.2238460070075234

foo_in              	0.05619414099783171
foo_casefold        	1.264593858999433
foo_ancor_match     	2.192337133004912
foo_plain_match     	2.2391775529977167
foo_plain_search    	2.3812111250008456
foo_ancor_search    	3.5944648580043577
```
The cal stuff are all matching search strings and the foo stuff are failed searches, I search the khard README.md, the results are sorted by time. We can already see here that the result depends on weather the search string matches or not. And we can also see that casefolding is expensive. I expect that the unicode normalisation mentioned in the SO thread linked above is also expensive.

Then again I don't know if this is the best option for optimizing the code for speed.  What are your findings? Where does khard spend most of its time when it feels slow to you?

**Comment by protist** [comment]
Sorry for the delay; I've been busy again.

> the fluctuation on your system seems to be quite high. Try to minimize load

I did try to do this at the time. I'm testing on a pretty good system (i7-7700, M.2 SSD), so I'm surprised that there's a large effect from load, but I'll alternate tests next time. I did do the last test on a different day, so maybe there was some batch effect involved.

> #167 is merged into develop so you can check that out. But I think it mostly affects `-f`.

I tested two rounds, alternating between branches. There was no video or anything like that playing. I tested with `time (for i in {1..10}; do khard ls -f foo bar > /dev/null; done)`. This actually matches with a single existing entry. The results are as follows.
```
Master:  2.780
Develop: 2.782
Master:  2.701
Develop: 2.834
```
So I still see no difference between master and develop. *However*, looking back through my previous posts, I can see that the base speed of the master branch has improved over time.

My first test on 21 Nov 2016 was `1.442 s`. Then, on a newer system on 18 Jan 2018, was `0.5320 s`, then 25 Jan 2018 was `0.3106 s`, and now it's `0.2780`. So it's much better now! But oddly enough, I've just never seen any improvement in master versus develop (or with the patch)!

> Where does khard spend most of its time when it feels slow to you?

Originally, I found it sluggish when querying from within mutt, i.e. when calling `khard email --parsable %s`. The original lookup was very slow, but now it's ~5-fold faster for me, it's much more useable. I'm not sure if more improvement could be made, although faster is always better!

**Comment by three-comrades** [comment]
If you really want to use casefolding (which seems to me more confusing than useful - `'ß'.casefold() -> ss`, but `'ä'.casefold() -> ä` instead of `ae`) and/or normalization, then maybe it's better to use some kind of persistent storage so that the expensive data transformation is decoupled from the search. 

[mate.rs](https://github.com/pimutils/mates.rs) simply creates a text file (which has to be updated/kept in sync manually) with some information from the vcards and then calls grep on it for email completion. That may be enough for most use cases, but it would be nice if the file would be kept automatically in sync.

**Comment by lucc** [comment]
I have uploaded my proof of concept for a caching address book class which I had lying around: https://github.com/lucc/khard/tree/feature/cached-abook  This is very much WIP but you can start to imagine ...

This implementation uses a concatenated vcard file but one could also implement caching via a sqlite db or I don't know what. Keeping things in sync would be easy I imagine. We just check the modification times on the cache and the originals and rebuild the cache when the originals are newer (think makefile). All modifying commands would work on the originals so they either don't search the cache (they don't need to be as fast anyways I think) or the cache has to keep track of the connection to the original files.

No idea if any of these would be faster than what we have now. As far as I could find out with cProfile our biggest speed bump comes from parsing vcard files. I am not sure that would change with the above example (especially as `-f` might be impossible to implement with this).

**Comment by GSI** [comment]
The one thing I keep wondering about is why not just use `grep` on the source files for searching?

All one needs to do is ignore anything before a colon. And search in the rest.

So if search term was '123' then something like `grep 1.*2.*3` does the trick.
If search term was '123' plus 'anna' then `grep 1.*2.*3` followed by `grep a.*n.*n.*a` on the reduced set of files get's you there.

Ofc it doesn't have to be command-line grep, but a corresponding python function. If lookup then still is slow, that would mean an issue with the function used as opposed to an issue to be tacklied in **khard** itself.

(Caching I think goes too far, since you can just pre-generate the necessary files for mutt - as suggested before and perfected by @protist in https://github.com/scheibler/khard/issues/103#issuecomment-270289046)

**Comment by lucc** [comment]
@GSI for simple cases `grep` (or any plain text or regex seach) will do. But the vcard format has some more details that might be interesting when searching with grep: there are continuation lines that do not start with a property name, so there might be lines without a colon. And there might be colons in values so there might be lines with more than one colon.

Luckily [version 3](https://tools.ietf.org/html/rfc2426) and [version 4](https://tools.ietf.org/html/rfc6350) are much simpler than version 2.1 (wich khard does not officially support. Things like quoted-printable encoded values, which would make grepping much harder,  have been removed for example.

But after finding the vcard we still have to parse it in order to properly understand and display it. IncidentallyG this is what we do with the `-f` option I think :)  First we open the file and search it with `re.search` and only parse the matching files. But as noted above I think we can still investigate if we can 
1. speed up the preselection search
2. improve the precision of the preselection
3. speed up `vobject`/vcard parsing (?) or somehow reduce it (see 2).

About simple external caching: As mentioned above this is what I currently use from my mail client and it works quite well: https://github.com/scheibler/khard/issues/103#issuecomment-269750576
One could obviously try to encode more information into the comment field of the mutt compatible line oriented format (like nickname, categories, roles, ...) but I have not tried that yet.

**Comment by three-comrades** [comment]
I have a SSD and I created a test vcard set of 5000 files.
khard 0.12.2:
```
./test.sh
time (for i in {1..10}; do khard list foo bar> /dev/null; done)
real	0m46,701s
user	0m44,520s
sys	0m2,181s
time (for i in {1..10}; do khard list -f foo bar > /dev/null; done)
real	0m5,691s
user	0m4,941s
sys	0m0,746s
```

The develop branch:
```
./test.sh 
time (for i in {1..10}; do khard list foo bar> /dev/null; done)
real	0m5,093s
user	0m4,561s
sys	0m0,526s
time (for i in {1..10}; do khard list -f foo bar > /dev/null; done)
real	0m5,044s
user	0m4,437s
sys	0m0,602s
```

tl;dr: on master it was pretty unusable without `-f`, now it takes around 500 ms per query.  This is great!

If we want to get into 'feels snappy' territory, the bazaar developers wrote a bit about optimizing the startup time of python programs: [click](https://files.bemusement.org/talks/OSDC2008-FastPython/). Some quick measurements with a simple patch that just returns, once we hit main:
```
diff --git a/khard/khard.py b/khard/khard.py
index 97af571..c8665d6 100644
--- a/khard/khard.py
+++ b/khard/khard.py
@@ -1661,6 +1661,7 @@ def parse_args(argv):
 
 
 def main(argv=sys.argv[1:]):
+    return
     args = parse_args(argv)
 
     # if args.action isn't one of the defined actions, it must be an alias
```

```
time (for i in {1..10}; do khard --help > /dev/null; done)

real	0m3,198s
user	0m2,914s
sys	0m0,284s
```
And simply loading the python interpreter as a baseline:
```
time (for i in {1..10}; do python -c exit > /dev/null; done)

real	0m0,196s
user	0m0,162s
sys	0m0,035s
```

With these non-scientific tests, the 500 ms per search query on my system are roughly:
 - 20 ms for starting the interpreter
 - around 300 ms until we hit `main`
 - so, maybe 200 ms for the search + parsing the command line and what not

So, my guess would be that your search optimization is already pretty good.

**Comment by lucc** [comment]
@three-comrades I am wondering why there is no difference between the normal search and the search with `-f` on develop. Otherwise interesting stuff about the import times.

So I will point out these things to all the people measuring the execution time of khard:
* please use a minimal config file (check `-c` or `$KHARD_CONFIG`) you can copy the minimal config file from test/fixtures/minimal.conf and modify the address books section.
* whichever config file you use, make sure that `[vcard] search_in_source_file = false`, as `true` would be the same as `-f`. The default (if nothing is set in the config file) is `false`.

**Comment by three-comrades** [comment]
>  I am wondering why there is no difference between the normal search and the search with -f on develop. 

I thought it was weird, too, but for what it's worth I get very similar times with a minimal config.

**Comment by lucc** [comment]
I will close this. Any new speed issues should be discussed in a new issue.

**Comment by protist** [comment]
Thanks @lucc. From my perspective, the improvements I mention [above](https://github.com/scheibler/khard/issues/103#issuecomment-362980228) make khard perfectly fine for my usage.
